lib/infomon.rb

```diff
@@ -0,0 +1,515 @@
+          hide_lines = true
+          nil
+        else
+          server_string
+        end
+      }
+      DownstreamHook.add('infomon_info', action)
+      echo 'checking stats...'
+      put 'info'
+      wait_until { done }
+
+      hide_lines = done = false
+      action = proc { |server_string|
+        if hide_lines
+          if server_string =~ %r{<output class=""/>|<prompt}
+            DownstreamHook.remove('infomon_skills')
+            done = true
+          end
+          nil
+        elsif server_string =~ %r{^\s*(?:<.*?>)?#{Char.name}(?:</a>)? \(at level}o
+          hide_lines = true
+          nil
+        else
+          server_string
+        end
+      }
+      DownstreamHook.add('infomon_skills', action)
+      echo 'checking skills...'
+      put 'skills'
+      wait_until { done }
+
+      hide_lines = done = false
+      action = proc { |server_string|
+        if hide_lines
+          if server_string =~ /<prompt/
+            DownstreamHook.remove('infomon_society')
+            done = true
+          end
+          nil
+        elsif server_string == "<pushBold/>\r\n"
+          hide_lines = true
+          nil
+        else
+          server_string
+        end
+      }
+      DownstreamHook.add('infomon_society', action)
+      echo 'checking society...'
+      put 'society'
+      wait_until { done }
+
+      done = false
+      action = proc { |server_string|
+        if server_string =~ /You currently have .*? citizenship in|You don't seem to have citizenship\./
+          DownstreamHook.remove('infomon_citizenship')
+          done = true
+          nil
+        else
+          server_string
+        end
+      }
+      DownstreamHook.add('infomon_citizenship', action)
+      echo 'checking citizenship...'
+      put 'citizenship'
+      wait_until { done }
+
+      silence_me
+
+    end
+
+    if $infomon_values['active_spells'].empty?
+      echo 'checking active spells...'
+      $process_legacy_spell_durations = true
+    end
+    echo 'done'
+
+    #
+    # Load spell timers
+    #
+    $infomon_values['active_spells'].each_pair do |spell_num, timeleft|
+      if (spell = Spell[spell_num.to_i])
+        if defined?(spell.real_time) && spell.real_time
+          timeleft = (timeleft - Time.now.to_f) / 60.0
+          if timeleft.positive?
+            spell.timeleft = timeleft
+            spell.active = true
+          end
+        else
+          spell.timeleft = timeleft
+          spell.active = true
+        end
+      else
+        echo "spell not loaded: #{spell_num}"
+      end
+    end
+    Spellsong.load_serialized = $infomon_values['Spellsong'] if $infomon_values['Spellsong']
+
+    #
+    # Register ;magic and ;banks commands moved
+    #
+
+    $infomon_sleeping = false
+    $infomon_bound = false
+    $infomon_silenced = false
+    $infomon_calmed = true
+    $infomon_cutthroat = false
+    #
+    # Save function
+    #
+    def self.save_proc
+      @save_value = $infomon_values
```

Comment:
I don't understand why we are copying a global var to an instance var like this

Reply:
May be a bit too conservative, but I was capturing the current global var state in case it tried to change on me mid write.  Same thinking as behind the $1.dup convention.

What I'd really like to do is have an instance variable throughout - but I clearly have yet to conquer namespace and scope.  I'd rather not have $infomon_values, instead @infomon_values.  Perhaps, in time.